Add s3_bucket_notification module

aljazkosir opened the pull request:
##### SUMMARY
PR to introduce `lambda_bucket_event` module.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
 - lib/ansible/modules/cloud/amazon/lambda_bucket_event.py
 - test/units/modules/cloud/amazon/test_lambda_bucket_event.py

##### ADDITIONAL INFORMATION
This module allows the management of AWS Lambda function bucket event mappings via the Ansible framework. Use module [lambda](https://docs.ansible.com/ansible/latest/modules/lambda_module.html) to manage the lambda function itself, [lambda_alias](https://docs.ansible.com/ansible/2.4/lambda_alias_module.html) to manage function aliases and [lambda_policy](https://docs.ansible.com/ansible/2.4/lambda_policy_module.html) to modify lambda permissions.

Example:
```
---
# Example that creates a lambda event notification for a bucket
- hosts: localhost
  gather_facts: no
  tasks:
  - name: Process jpg image
    s3_bucket_notification:
      state: present
      event_name: on_jpg_add_or_remove
      bucket_name: test-bucket
      function_name: arn:aws:lambda:us-east-2:525810310200:function:test-lambda
      events: ["s3:ObjectCreated:*", "s3:ObjectRemoved:*"]
      prefix: images/
      suffix: .jpg
```

aljazkosir committed:
Add lambda_bucket_event module

aljazkosir commented:
bot_status

aljazkosir commented:
ready_for_review

aljazkosir commented:
ready_for_review

willthames commented:
This module seems the wrong approach - I'd prefer that this was just a parameter in the `s3_bucket` module

mattclay commented:
The policies used by our CI system will need to be updated:

https://github.com/mattclay/aws-terminator/tree/master/aws/policy

If the permissions added only add read-only actions, then only policy updates are needed. If the policy changes allow creation of resources then we'll need to make additional changes.

aljazkosir commented:
@mattclay, we require both `GetBucketNotificationConfiguration` and `PutBucketNotificationConfiguration`.

mattclay commented:
@aljazkosir That should be OK then, since the notification configuration should be removed when the bucket it's attached to is removed.

miha-plesko commented:
Hi @willthames thanks for bringing up the question whether we should do a separate module or extend the existing one. It may be my personal opinion but isn't it better to have every module focused on a single task? Like one module to "ensure bucket exists" and another one to "ensure bucket connected to lambda".

There are at least these drawbacks if we stuff the lambda connectivity to the existing `s3_bucket` module:
- `s3_bucket` parameter list would become a mess because there will be 4 additional parameters only related to lambda connectivity i.e. "conditionally requred"
- it would become impossible to connect lambda to existing bucket without modifying the existing bucket
- if task would yield `changed: yes` then you don't know whether the bucket itself was updated, or just lambda connectivity
- if task would raise then you don't know whether the bucket itself is erroring or just lambda connectivity
- it would become more cumbersome to test the module because it would be doing yet another thing

Now, I may be wrong and am looking forward to your opinion. Its always difficult to determine how much a module is supposed to cover and when it's time for a new module.

/cc @tadeboro with whom I had similar discussion few months ago. I'd value your opinion on the matter as well

samdoran commented:
I have update the S3 bucket policies and restarted the failed tests.

tadeboro commented:
If I put myself in the shoes of the playbook author, I would rather see this functionality as a separate module.

Why? If we read through the documentation for the `s3_bucket` module, it indicates that `s3_bucket`module's purpose is to create or delete a bucket. But establishing a link between a bucket and a lambda function is not one of those things that we would only do at a bucket creation time.

Another thing that I find quite important is playbook readability. If I have a separate module for establishing a link between different components, I can write my playbooks like this:

```
- name: Get or create a bucket
  s3_bucket:
    ...

- name: Get or create a lambda function
  lambda:
    ...

- name: Link a bucket and a lambda function
  lambda_bucket_event:
    ...
```

Things read nicely, and at each step, we do one and one thing only. If we add this functionality to the `s3_bucket` module, the playbook will look like this:

```
- name: Get or create a lambda function
  lambda:
    ...

- name: Get or create a bucket and link it to a lambda function
  s3_bucket:
    ...
```

Notice that "ugly" **and** in the description of the second task? A clear indicator that the module in question is doing too much.

Just my two cents.

jillr commented on lib/ansible/modules/cloud/amazon/lambda_bucket_event.py:
```diff
@@ -0,0 +1,255 @@
+  alias:
+    description:
+      - Name of the function alias. Mutually exclusive with C(version).
+    required: true
```
```suggestion
    required: false
```
The arg_spec is correct for this

jillr commented on lib/ansible/modules/cloud/amazon/lambda_bucket_event.py:
```diff
@@ -0,0 +1,255 @@
+'''
+
+RETURN = '''
+'''
```
Would like to see this documented

aljazkosir replied:
Our module is not returning anything in particular, hence the empty return.

willthames replied:
It should though. It should return the current set of notifications.

jillr reviewed (commented):
Thanks @miha-plesko and @tadeboro , I agree with the perspective you've added here.

I'd like to see more integration tests though.  Could you please add additional integration tests covering:
* Check mode
* Idempotency on CRUD operations
* Testing of mutually exclusive options

willthames commented:
@tadeboro thanks for laying out that reasoning, definitely helpful to me.

I still don't agree with the name of the module though, it probably should be s3_bucket_notifications or something, and lambda notifications should just be a special case (we don't necessarily need to implement all the other cases right now though)

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.put_bucket_notification_configuration

tadeboro commented:
> I still don't agree with the name of the module though, it probably should be s3_bucket_notifications or something, and lambda notifications should just be a special case (we don't necessarily need to implement all the other cases right now though)

I did not comment on the module name because I am not too intimately familiar with the AWS architecture/nomenclature (and was too lazy to go and look it up). But your suggestion is a significant improvement over the current proposal, and I am all-in for replacing the name. Maybe we should use a singular form to be a bit more consistent with the rest of the Ansible modules?

@willthames Thank you for helping us improve this module. Your input is greatly appreciated.

willthames commented:
@tadeboro thanks for that! You're quite right, it should indeed be singular: https://github.com/ansible/proposals/issues/10

aljazkosir commented:
Thank you guys, I added a few more integration test and changed the name of the module.

@willthames @jillr should we add ourselves to `BOTMETA.yml` as module maintainers?

willthames commented:
@aljazkosir you shouldn't need to as long as you're in the authors list - botmeta is only needed for overriding that.

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,266 @@
+  lambda_function_arn:
+    description:
+      - The ARN of the lambda function.
+    required: true
```
Looking at this from a future proofing point of view, I'm not sure `required` should be true here (that also helps if you just want to delete a notification with only `event_name` and `state`)

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,266 @@
+        lambda_function_arn=dict(default=None, aliases=['function_arn']),
+        bucket_name=dict(required=True),
+        events=dict(type='list', default=[], choices=event_types),
+        prefix=dict(required=False, default=''),
```
required=False is the default

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,266 @@
+        events=dict(type='list', default=[], choices=event_types),
+        prefix=dict(required=False, default=''),
+        suffix=dict(required=False, default=''),
+        alias=dict(required=False, default=None),
```
default=None is the default

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,266 @@
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        mutually_exclusive=[['alias', 'version']],
+        required_if=[['state', 'present', ['events', 'lambda_function_arn']]]
```
I'd like to remove `lambda_function_arn` in case anyone ever wants to implement other notification types

willthames commented on test/integration/targets/aws_s3_bucket_notification/tasks/main.yml:
```diff
@@ -0,0 +1,316 @@
+  # ============================================================
+  # Test idempotency of CRUD
+
+    - name: test that configuration is the same as previous task
```
I'd like to see an idempotency check that order of events doesn't matter

aljazkosir commented:
@willthames I updated the arguments and added test that checks that order of events doesn't matter

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,267 @@
+    description:
+      - Name of the function alias. Mutually exclusive with C(version).
+    required: false
+    aliases: ['lambda_alias']
```
Can we just call this argument `lambda_alias` without the alias (just thinking about future proofing for other notification targets)

willthames commented on lib/ansible/modules/cloud/amazon/s3_bucket_notification.py:
```diff
@@ -0,0 +1,267 @@
+    required: false
+    aliases: ['lambda_alias']
+    type: str
+  version:
```
Can we just call this argument `lambda_version` without the alias (just thinking about future proofing for other notification targets)

willthames commented on test/integration/targets/aws_s3_bucket_notification/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/aws
```
Need to make the test suite name match the module name. I don't mind if that's adding `aws_` to the module name or removing `aws_` from the test suite name

aljazkosir replied:
Removing `_aws` from the test suite name seems better to me.

willthames reviewed (approved):
I've tested this locally, everything looks great. Good to merge from my point of view

jillr reviewed (changes requested):
Overall this looks good, but I'm concerned about the 41MB function.zip.  We have to carry this around in the repo (and ansible is already fairly hefty), and I'm getting timeouts on the "register lambda" test some of the time. This will likely lead to these tests being unstable in CI.  It would be better to use a simple hello world style script like `test/integration/targets/aws_lambda/files/mini_lambda.py` or `test/integration/targets/lambda_policy/files/mini_http_lambda.py`.

jillr commented:
I'd also personally prefer it if the function were in-repo as a regular file, then zipped in the test, from a review and accounting perspective.  It makes it much easier to track and ensure that nothing is sneaking into the repo via a binary.

mattclay commented on test/integration/targets/s3_bucket_notification/tasks/main.yml:
```diff
@@ -0,0 +1,329 @@
+    - name: move lambda into place for archive module
+      copy:
+        src: "function.zip"
+        dest: "{{output_dir}}/function.zip"
```
We should not include a file of this size in the repo.  Use a very tiny lambda function and create the zip file during the test. If that's not possible, then we can put the zip file in our S3 bucket for CI -- but lets avoid that if possible.

mattclay replied:
Ideally the lambda would be a singe python file with no requirements outside the standard library, then no zip would be needed. Keeping it python vs another language makes it easier to review since most contributors will be familiar with python.

aljazkosir commented:
@jillr @mattclay Of course, I totally forgot about that zip, sorry about that. It is now zipped during the test without any extra requirements, so it should be fairly small. 
ready_for_review

aljazkosir commented:
I changed the authors, so the right people will be pinged.

jillr reviewed (approved):
Thanks @aljazkosir that looks much better, appreciate all your work on this module!
